File: 2026/07-July/02-Thursday/solution.java
import java.util.*;

class Solution {
    public boolean findSafeWalk(List<List<Integer>> grid, int health) {
        int m = grid.size();
        int n = grid.get(0).size();
        
        // Directions for moving up, down, left, right
        int[][] dirs = {{0, 1}, {1, 0}, {0, -1}, {-1, 0}};
        
        // dist[i][j] stores the minimum health lost to reach cell (i, j)
        int[][] dist = new int[m][n];
        for (int[] row : dist) Arrays.fill(row, Integer.MAX_VALUE);
        
        Deque<int[]> deque = new ArrayDeque<>();
        
        // Initialize starting point
        dist[0][0] = grid.get(0).get(0);
        deque.offerFirst(new int[]{0, 0});
        
        while (!deque.isEmpty()) {
            int[] curr = deque.pollFirst();
            int r = curr[0], c = curr[1];
            
            // If we reached the bottom-right corner
            if (r == m - 1 && c == n - 1) {
                return health - dist[r][c] > 0;
            }
            
            for (int[] dir : dirs) {
                int nr = r + dir[0];
                int nc = c + dir[1];
                
                if (nr >= 0 && nr < m && nc >= 0 && nc < n) {
                    int weight = grid.get(nr).get(nc);
                    
                    if (dist[r][c] + weight < dist[nr][nc]) {
                        dist[nr][nc] = dist[r][c] + weight;
                        
                        // 0-1 BFS logic
                        if (weight == 0) {
                            deque.offerFirst(new int[]{nr, nc});
                        } else {
                            deque.offerLast(new int[]{nr, nc});
                        }
                    }
                }
            }
        }
        
        return false;
    }
}

File: 2026/07-July/02-Thursday/solution.kt
import java.util.ArrayDeque

class Solution {
    fun findSafeWalk(grid: List<List<Int>>, health: Int): Boolean {
        val m = grid.size
        val n = grid[0].size
        val dist = Array(m) { IntArray(n) { Int.MAX_VALUE } }
        val dirs = arrayOf(intArrayOf(0, 1), intArrayOf(1, 0), intArrayOf(0, -1), intArrayOf(-1, 0))
        
        val deque = ArrayDeque<IntArray>()
        dist[0][0] = grid[0][0]
        deque.offerFirst(intArrayOf(0, 0))
        
        while (deque.isNotEmpty()) {
            val curr = deque.pollFirst()!!
            val r = curr[0]
            val c = curr[1]
            
            if (r == m - 1 && c == n - 1) return health - dist[r][c] > 0
            
            for (d in dirs) {
                val nr = r + d[0]
                val nc = c + d[1]
                if (nr in 0 until m && nc in 0 until n) {
                    val w = grid[nr][nc]
                    if (dist[r][c] + w < dist[nr][nc]) {
                        dist[nr][nc] = dist[r][c] + w
                        if (w == 0) deque.offerFirst(intArrayOf(nr, nc))
                        else deque.offerLast(intArrayOf(nr, nc))
                    }
                }
            }
        }
        return false
    }
}
